x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -851,14 +852,16 @@ Collection<Object> createComponents(
 
         // realms construction
         final NativeUsersStore nativeUsersStore = new NativeUsersStore(settings, client, systemIndices.getMainIndexManager());
+        final ClusterStateRoleMapper clusterStateRoleMapper = new ClusterStateRoleMapper(settings, scriptService, clusterService);
         final NativeRoleMappingStore nativeRoleMappingStore = new NativeRoleMappingStore(
             settings,
             client,
             systemIndices.getMainIndexManager(),
-            scriptService
+            scriptService,
+            new ReservedRoleMappings(clusterStateRoleMapper)
         );
-        final ClusterStateRoleMapper clusterStateRoleMapper = new ClusterStateRoleMapper(settings, scriptService, clusterService);
-        final UserRoleMapper userRoleMapper = new CompositeRoleMapper(nativeRoleMappingStore, clusterStateRoleMapper);
```

Comment:
This change (getting rid of `CompositeRoleMapper`) is a little nuanced. The reason I'm making it is:

The PR changes the `NativeRoleMappingStore` to fetch "reserved" (i.e., cluster-state-based ones) role mappings and use these instead of any native role mappings with the same name. This means that we read cluster-state role mappings _in_ the `NativeRoleMappingStore`, and substitute mappings based on what we read. If we stuck with the `CompositeRoleMapper` here, we'd "double-read" cluster-state-based role mappings. 

An alternative I originally considered is to filter out native role mappings that match reserved ones and then simply rely on `CompositeRoleMapper` here to fill in cluster-state ones. However, this would introduce a race condition:

1. When we get role mappings from the native store, we read cluster-state and filter out.
2. We returning the result.
3. Concurrently, cluster state changes (mappings added or removed).
4. We read cluster state again, including the new mappings.

However, this means we have a "stale" view obtained in step 2.

By going with current approach in this PR instead (remove `CompositeRoleMapper`), we avoid this race condition since we only read cluster state once.

Reply:
Tricky! I think your change here looks good. I thought of an alternative explained below. 

I think it would be clearer to keep the role mapping stores separate and then leave the merging of them to some external class. Priorities can change and the way we want to merge them could change too. With your change `ClusterStateRoleMapper` is injected into the `NativeRoleMappingStore` through the wrapping class `ReservedRoleMappings` and you're using `ClusterStateRoleMapper` "internals" (`getMappings` - not exactly an internal but I think it was only public for testing purposes). It does violate the `UserRoleMapper` interface a little. 

I think adding a new `MergingCompositeRoleMapper` (or a better name) that handles both read and write (so both `getRoleMappings` and `putRoleMapping`) would be more general. For read it would merge the result of the provided `UserRoleMapper`s where priority could be the order of arguments passed (so `ClusterStateRoleMapper` would be passed first to have priority). For writes it could check the `ClusterStateRoleMapper` for reserved names before writing (according to the configuration) to the native store.

Reply:
I've considered this (see a stub of it [here](https://github.com/elastic/elasticsearch/pull/113853/files)) however it comes out a much bigger change, and I'm not sure it's worth it because the abstraction itself is sort of neither here nor there. The problems are:

The new class is neither Merging, nor a role mapper. Instead it's a class that merges on reads, but prevents reserved writes. So it's more of a `ReservedRoleMappingWrappedStore`. And it's not just a role mapper (that only handles resolving roles) but a role mapping store. The latter requires a whole new interface or [abstract class](https://github.com/elastic/elasticsearch/pull/113853/files#diff-43395b0d3ac8022699957e43079d534bdf8b787fea686b1b3a4677a3f10b7278R18), and plugging that interface in everywhere `NativeRoleMappingStore` is currently used. Do-able, but `NativeRoleMappingStore` is used in several actions and tests. 

I'm just not sure that it's worth it to introduce all this generic structure to handle something that's unlikely to generalize. WDYT?

Reply:
Yes, I see what you mean. The interfaces are note exactly built for that and the change is big. 

>I'm just not sure that it's worth it to introduce all this generic structure to handle something that's unlikely to generalize. WDYT?

I think the coupling this introduces is a little concerning, but I don't think this is the right time to be making a huge refactor, so I think leaving it is fine.